pandas/_libs/src/parser/tokenizer.c

```diff
@@ -1774,11 +1774,16 @@ double precise_xstrtod(const char *str, char **endptr, char decimal,
 
 double round_trip(const char *p, char **q, char decimal, char sci, char tsep,
                   int skip_trailing, int *error, int *maybe_int) {
+    PyGILState_STATE gstate;
```

Comment:
this is likely *very* expensive.

Reply:
is this what cython code actually generates with a nogil?

Reply:
This code ensures The GIL, `nogil` by contrast would release it.

Below is what master generates for the nogil funcs:

```c
    /* "pandas/_libs/parsers.pyx":1667
 *     na_fset = kset_float64_from_list(na_flist)
 *     if parser.double_converter_nogil != NULL:  # if it can run without the GIL
 *         with nogil:             # <<<<<<<<<<<<<<
 *             error = _try_double_nogil(parser, parser.double_converter_nogil,
 *                                       col, line_start, line_end,
 */
    {
        #ifdef WITH_THREAD
        PyThreadState *_save;
        Py_UNBLOCK_THREADS
        __Pyx_FastGIL_Remember();
        #endif
        /*try:*/ {

          /* "pandas/_libs/parsers.pyx":1668
 *     if parser.double_converter_nogil != NULL:  # if it can run without the GIL
 *         with nogil:
 *             error = _try_double_nogil(parser, parser.double_converter_nogil,             # <<<<<<<<<<<<<<
 *                                       col, line_start, line_end,
 *                                       na_filter, na_hashset, use_na_flist,
 */
          __pyx_v_error = __pyx_f_6pandas_5_libs_7parsers__try_double_nogil(__pyx_v_parser, __pyx_v_parser->double_converter_nogil, __pyx_v_col, __pyx_v_line_start, __pyx_v_line_end, __pyx_v_na_filter, __pyx_v_na_hashset, __pyx_v_use_na_flist, __pyx_v_na_fset, __pyx_v_NA, __pyx_v_data, (&__pyx_v_na_count));
        }
```

For simplicity this stays in that block but just ensures The GIL before actually doing anything.

So I see your point. An alternate strategy here is to provide a flag within this function if its safe to release The GIL rather than releasing then reacquiring. Let me see what that looks like

Reply:
hmmm that's not what i mean. why don't we just mark this function as nogil? seems a lot simpler

Reply:
Ah...well the function actually interacts with Python objects so it does require the GIL

Reply:
Though could maybe refactor to not use Python objects (really the Exception handling items are the Python objects) if that is preferable

Reply:
Ref original issue #15140